### DISCUSSION TITLE
Rule don't apply on multi relation fields

### DISCUSSION THREAD

**Discussion by iskaa02** [pocketbase/pocketbase#788] [Q&A]
Create or Update rules don't apply to a field with multiple relation field
take this rule as an example`a.c=a.b.c` the b field takes multiple `c` records as a value, when there's one id provided the rule work as expected but when c has multiple record ids the rule don't.

**Comment by ganigeorgiev** [comment]
@iskaa02 This is because you are comparing the entire json array. Assuming that both `c` are multi-relation fields, you can compare the relation ids:

```
a.c.id = a.b.c.id
``` 
The above rule will be true if there is at least one matching record.

_I'll move the issue to Q&A discussion._

**Comment by iskaa02** [comment]
Sorry the example i provided wasn't detailed enough, I have done what you proposed, the issue is that i thought the rule will be false if there's one id that don't match, and I think that is the expected behavior.

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand. Let's say that `a.c` has value `['id1', 'id2']` and `a.b.c` has value `['id1', 'id3']`.

Comparing `a.c = a.b.c` should result to `false` because the 2 json string representations are different.

Comparing `a.c.id = a.b.c.id` should result to `true` (this is similar to the sql `IN`). This is also consistent with for example: `a.c.someOtherField > 10` (`true` if `c` has at least one record with `someOtherField > 10`).

The following should also result to `true` - `'id1' = a.b.c.id` (_the order of the arguments doesn't matter_)

**Reply by iskaa02** [reply]
I think the abstract examples made it a little bit complex, Sorry about that.
I have 3 relations `product` and `product_options` and `orders`

### Product
we care only about the ID field in this table 

### Product_options
```
+---------+---------------------------+
|  name   |           type            |
+---------+---------------------------+
| product | relation to table product |
+---------+---------------------------+

```
### Orders
```
+---------+-----------------------------------------------+
|  name   |                     type                      |
+---------+-----------------------------------------------+
| product | relation to table product                     |
| option  | one-to-many relation to table product_options |
+---------+-----------------------------------------------+
```
The API rule in create and update for the orders table  is `option.product.id=product.id`. it should check if all the options belong to the provided product, I thought it was a bug, not sure if it is the intended behavior, I already solved this problem using hooks but i think The API rule are a bit limited at the moment.

**Reply by ganigeorgiev** [reply]
This is an intended behavior and as mentioned earlier we need only one record from the milti-relation conditions to match in order to assume that the rule is `true`. This is because internally the rule is converted to a left join and you end up with a query similar to:

```sql
# pseudo-sql (the actual query is a little more complicated)
SELECT DISTINCT orders.*
FROM orders
LEFT JOIN orders.options as orders_options_each # join each json array option
LEFT JOIN product_options ON product_options.id = orders_options_each.value
WHERE (
    orders.id='RECORD_ID' AND
    product_options.product = orders.product
)
LIMIT 1
```

On the other hand, I understand your use case and I see why you or other users could assume that the rule is applied on every relation field. Or in other words, I guess what you want is a query similar to:

```sql
# pseudo-sql
SELECT orders.*
FROM orders
LEFT JOIN orders.options as orders_options_each # join each json array option
LEFT JOIN product_options ON product_options.id = orders_options_each.value
WHERE orders.id='RECORD_ID'
GROUP BY orders.id
HAVING product_options.product = orders.product # or maybe count the non-null product_options.product
LIMIT 1
```

Personally, the current behavior is what I would expect, but maybe this is only because I'm familiar with the codebase and I know how the rule will be translated. I have added the request to my todo to check it later together with the `IN` operator for the plain multi-select options field.

**Reply by ganigeorgiev** [reply]
@iskaa02 _Just to let you know that multi-match/match-all conditions for multi-relation fields (both via the `@request.data.someRel.*` and `someRel.*`) are implemented in [v0.11.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.11.0)._
